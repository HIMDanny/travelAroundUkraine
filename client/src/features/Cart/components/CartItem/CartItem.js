// React
import React, { useState, useEffect } from 'react';
import { useNavigate } from 'react-router-dom';
import PropTypes from 'prop-types';

// Redux
import { useDispatch } from 'react-redux';

// MUI components
import {
  styled,
  Stack,
  CardActions,
  CardMedia,
  Button,
  Typography,
  CardContent as MuiCardContent,
  IconButton,
  TextField,
} from '@mui/material';
import { AccessTime, Remove, Add, DeleteOutline } from '@mui/icons-material';

// Components
import { ReactComponent as CoinsIcon } from '../../../../assets/svg/CoinsIcon.svg';
import { AlertModal } from '../../../../components';

// Redux Thunk functions
import {
  addProduct,
  addProductToLocal,
  decreaseQuantity,
  deleteProduct,
  decreaseProductFromLocal,
  delProductFromLocal,
} from '../../../../store/slices/cartSlice/cartSlice';
import getProduct from '../../../../api/getProduct';

const CardContainer = styled(Stack)(({ theme }) => ({
  position: 'relative',
  backgroundColor: theme.palette.primary.contrastText,
  borderRadius: 10,
  boxShadow: '0px 0px 40px rgba(0, 0, 0, 0.05)',
}));

const CardContent = styled(MuiCardContent)(({ theme }) => ({
  [theme.breakpoints.up('tablet')]: {
    paddingTop: 24,
    width: '100%',
  },
}));

const CardImage = styled(CardMedia)(({ theme }) => ({
  width: '100%',
  maxHeight: 240,
  borderRadius: '10px 10px 0 0',

  [theme.breakpoints.up('tablet')]: {
    width: 238,
    height: 'auto',
    borderRadius: '10px 0 0 10px',
  },
}));

const CardTitle = styled((props) => <Typography variant="h3" {...props} />)(({ theme }) => ({
  color: theme.palette.text.primary,
  marginBottom: 20,
  [theme.breakpoints.up('tablet')]: {
    marginRight: 45,
  },
}));

const CardButton = styled(Button)(({ theme }) => ({
  padding: '12px 85px',
  textAlign: 'center',

  [theme.breakpoints.up('tablet')]: {
    paddingInline: '43px',
  },
}));

const DeleteButton = styled((props) => (
  <IconButton color="error" {...props}>
    <DeleteOutline />
  </IconButton>
))(({ theme }) => ({
  position: 'absolute',
  top: 15,
  right: 15,

  [theme.breakpoints.down('tablet')]: {
    backgroundColor: theme.palette.common.white,
    transition: theme.transitions.create('background-color'),

    '&:hover': {
      backgroundColor: theme.palette.grey[300],
    },
  },
}));

const AmountField = styled((props) => <TextField size="small" autoComplete="off" {...props} />)({
  maxWidth: '56px',

  '& input': {
    textAlign: 'center',
  },
});

const CartItem = ({ imageUrls, name, currentPrice, duration, cartQuantity, itemNo, id, isLogin }) => {
  const navigate = useNavigate();
  const dispatch = useDispatch();
  const [isDeleteDialogOpen, setDeleteDialogOpen] = useState(false);

  const [amountError, setAmountError] = useState(false);

  useEffect(() => {
    // ? should error text be provided?
    // add error if amount less than 1
    if (cartQuantity < 1) {
      setAmountError(true);
    } else {
      setAmountError(false);
    }
  }, [cartQuantity]);

  const openDeleteDialog = () => setDeleteDialogOpen(true);

  const handleDeleteDialogClose = () => setDeleteDialogOpen(false);

  const handleDeleteFromCart = () => {
    isLogin ? dispatch(deleteProduct(id)) : dispatch(delProductFromLocal(id));

    setDeleteDialogOpen(false);
  };

  const increaseBtnHandler = async () => {
    if (isLogin) {
      dispatch(addProduct(id));
    } else {
      const product = await getProduct(itemNo);
      dispatch(addProductToLocal(product));
    }
  };

  const decreaseBtnHandler = () => {
    isLogin ? dispatch(decreaseQuantity(id)) : dispatch(decreaseProductFromLocal(id));
  };

  if (!imageUrls || !name || !currentPrice || !duration || !cartQuantity || !itemNo || !id) return null;

  return (
    <>
      <CardContainer direction={{ xs: 'column', tablet: 'row' }}>
        <DeleteButton onClick={openDeleteDialog} />
        {imageUrls ? <CardImage component="img" image={imageUrls[1]} alt="tour photo" /> : null}

        <CardContent>
          <CardTitle>{name}</CardTitle>

          <Stack direction="row" gap={1} alignItems="start" sx={{ marginBottom: '10px' }}>
            <Stack direction="row" gap={1} alignItems="center" flexBasis="100px">
              <CoinsIcon />
              <Typography variant="h3" component="span">
                {currentPrice * cartQuantity} â‚¬
              </Typography>
            </Stack>

            <Stack direction="row" gap={1} alignItems="center">
              <AccessTime color="primary" sx={{ height: '21px' }} />
              <Typography variant="h3" component="span">
                {duration}
              </Typography>
            </Stack>
          </Stack>

          <Stack direction="row" marginY="5px">
            <IconButton disabled={cartQuantity <= 1} onClick={decreaseBtnHandler}>
              <Remove />
            </IconButton>
            <AmountField value={cartQuantity} error={amountError} />
            <IconButton onClick={increaseBtnHandler}>
              <Add />
            </IconButton>
          </Stack>
          <CardActions>
            <CardButton onClick={() => navigate(`/tour/${itemNo}`)}>More details</CardButton>
          </CardActions>
        </CardContent>
      </CardContainer>
      <AlertModal
        open={isDeleteDialogOpen}
        onClose={handleDeleteDialogClose}
        onSubmit={handleDeleteFromCart}
        title="Delete this tour?"
        submitButtonText="Delete"
      >
        Are you sure you want to delete tour:
        <Typography component="span" fontStyle="italic">
          {name}
        </Typography>
      </AlertModal>
    </>
  );
};

CartItem.propTypes = {
  imageUrls: PropTypes.array,
  name: PropTypes.string,
  currentPrice: PropTypes.number,
  duration: PropTypes.string,
  cartQuantity: PropTypes.number,
  itemNo: PropTypes.string,
  id: PropTypes.string,
  isLogin: PropTypes.bool,
};

CartItem.defaultProps = {
  imageUrls: [],
  name: 'unknown',
  currentPrice: 0,
  duration: '',
  cartQuantity: 0,
  itemNo: '',
  id: '',
  isLogin: false,
};

export default CartItem;
