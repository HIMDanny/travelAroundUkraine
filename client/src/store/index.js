import { configureStore } from '@reduxjs/toolkit';
import userReducer from './slices/userSlice/userSlice';
import tourReducer from './slices/tourSlice/tourSlice';
import catalogueReducer from './slices/catalogueSlice/catalogueSlice';
import cartReducer from './slices/cartSlice/cartSlice';
import orderSlice from './slices/orderSlice/orderSlice';
import inFavoritesReducer from './slices/inFavoritesSlice/inFavoritesSlice';

const store = configureStore({
  reducer: {
    userReducer,
    tour: tourReducer,
    catalogue: catalogueReducer,
    cart: cartReducer,
    order: orderSlice,
    favorites: inFavoritesReducer,
  },
});

export default store;
