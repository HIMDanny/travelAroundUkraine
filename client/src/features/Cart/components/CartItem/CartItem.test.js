/* eslint-disable react/prop-types */
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import store from '../../../../store';
import CartItem from './CartItem';

jest.mock('@mui/material/Stack', () => ({ children }) => <div>{children}</div>);
jest.mock('@mui/material/CardActions', () => ({ children }) => <div>{children}</div>);
jest.mock('@mui/material/CardMedia', () => ({ children }) => <div>{children}</div>);
jest.mock('@mui/material/Button', () => ({ children }) => <button type="button">{children}</button>);
jest.mock('@mui/material/Typography', () => ({ children }) => <p>{children}</p>);
jest.mock('@mui/material/CardContent', () => ({ children }) => <div>{children}</div>);
jest.mock('@mui/material/IconButton', () => ({ children }) => <span>{children}</span>);
jest.mock('@mui/material/TextField', () => ({ children }) => <div>{children}</div>);

describe('CartItem snapshot test', () => {
  test('should CartItem match snapshot', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <MemoryRouter>
          <CartItem />
        </MemoryRouter>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
