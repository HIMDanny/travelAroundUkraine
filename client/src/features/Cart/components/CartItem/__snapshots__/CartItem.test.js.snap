// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItem snapshot test should CartItem match snapshot 1`] = `
<DocumentFragment>
  <div>
    <span>
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="DeleteOutlineIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM8 9h8v10H8V9zm7.5-5-1-1h-5l-1 1H5v2h14V4z"
        />
      </svg>
    </span>
    <div />
    <div>
      <p>
        Sightseeing tour of Chernivtsi
      </p>
      <div>
        <div>
          <svg>
            CoinsIcon.svg
          </svg>
          <p>
            69 â‚¬
          </p>
        </div>
        <div>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-1fb58gc-MuiSvgIcon-root"
            data-testid="AccessTimeIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
            />
            <path
              d="M12.5 7H11v6l5.25 3.15.75-1.23-4.5-2.67z"
            />
          </svg>
          <p>
            3 hours
          </p>
        </div>
      </div>
      <div>
        <span>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="RemoveIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13H5v-2h14v2z"
            />
          </svg>
        </span>
        <div />
        <span>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="AddIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
          </svg>
        </span>
      </div>
      <div>
        <button
          type="button"
        >
          More details
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
